src/main/java/org/zalando/nakadi/service/ConsumerLimitingService.java

```diff
@@ -0,0 +1,175 @@
+package org.zalando.nakadi.service;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.echocat.jomon.runtime.concurrent.RetryForSpecifiedCountStrategy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.zalando.nakadi.exceptions.ConnectionSlotOccupiedException;
+import org.zalando.nakadi.exceptions.NakadiRuntimeException;
+import org.zalando.nakadi.exceptions.NoConnectionSlotsException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
+import org.zalando.nakadi.repository.zookeeper.ZooKeeperHolder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+import static java.text.MessageFormat.format;
+import static org.echocat.jomon.runtime.concurrent.Retryer.executeWithRetry;
+
+@Service
+public class ConsumerLimitingService {
+
+    public static final String CONNECTIONS_ZK_PATH = "/nakadi/consumers/connections";
+
+    private static final Logger LOG = LoggerFactory.getLogger(CursorsService.class);
+    private static final List<String> SLOT_NAMES = ImmutableList.of("a", "b", "c", "d", "e");
+    private static final String ERROR_MSG = "You exceeded the maximum number of simultaneous connections to a single " +
+            "partition for event type '{0}', partition(s): {1}; max limit is {2} connections per client";
+
+    private final ZooKeeperHolder zkHolder;
+    private final int maxConnections;
+
+    @Autowired
+    public ConsumerLimitingService(final ZooKeeperHolder zkHolder,
+                                   @Value("${nakadi.stream.maxConnections}") final int maxConnections) {
+        this.zkHolder = zkHolder;
+        this.maxConnections = maxConnections;
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<ConnectionSlot> acquireConnectionSlots(final String client, final String eventType,
+                                                       final List<String> partitions)
+            throws NoConnectionSlotsException, ServiceUnavailableException {
+
+        final List<Optional<ConnectionSlot>> slots = new ArrayList<>();
+        try {
+            // acquire slots for all partitions
+            for (final String partition : partitions) {
+                Optional<ConnectionSlot> connectionSlot = Optional.empty();
+                try {
+                    connectionSlot = executeWithRetry(
```

Comment:
As you see the implementation doesn't use locks - because that puts additional load on ZK and also lock nodes are not cleaned after. So to avoid of creation of more than 5 connections the implementation uses named slots ("1", "2", "3", "4", "5"). And the possible situation is that two instances will try to capture the same free slot at the same time. In that case we should retry to try capture another free slot. 
(Maybe not the best explanation but feel free to approach me so that I can describe how it works on board)